modules/bclconvert/main.nf

```diff
@@ -0,0 +1,84 @@
+process BCLCONVERT {
+    tag 'demultiplexing'
+    label 'process_high'
+
+    if (params.enable_conda) {
+        exit 1, "Conda environments cannot be used when using bcl-convert. Please use docker or singularity containers."
+    }
+    container "nfcore/bclconvert:3.9.3"
+
+    input:
+    path samplesheet
+    path run_dir
+
+    output:
+    path "*.fastq.gz"               ,emit: fastq
+    path "Reports/*.{csv,xml,bin}"  ,emit: reports
+    path "Logs/*.{log,txt}"         ,emit: logs
+    path "InterOp/*.bin"            ,emit: interop
+    path "versions.yml"             ,emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+
+    """
+    bcl-convert \
+        $args \\
+        --output-directory $PWD \\
+        --bcl-input-directory ${run_dir} \\
+        --sample-sheet ${samplesheet} \\
+        --bcl-num-parallel-tiles ${task.cpus / 4} \\
+        --bcl-num-conversion-threads ${task.cpus / 4} \\
+        --bcl-num-compression-threads ${task.cpus / 4} \\
+        --bcl-num-decompression-threads ${task.cpus / 4}
```

Comment:
Is the reason for dividing the cores up documented somewhere?

Reply:
Yeah, there's nothing in the commit message https://github.com/nf-core/modules/commit/c426128325e5afa14dca0664ecca911dd76aae0a

Reply:
:confused:

Reply:
https://nf-co.re/docs/contributing/modules#resource-requirements point 3

This might be a good one for @Midnighter to add to his Structizer one... there is some discussion **somewhere** on slack where this was originally brought up :sweat_smile:

Reply:
Not sure it's such a good fit for an ADR. Those are more supposed to answer why certain design decisions were made. Seems well placed as it is in the docs you linked to.

Reply:
Resource requirements for multiple tools are documented here: https://nf-co.re/contributing/modules?q=resource#resource-requirements

Reply:
Here is at least part of the discussion, if anyone is interested: https://nfcore.slack.com/archives/CJRH30T6V/p1632822087211400

Reply:
I just meant this specific instance of this one, and why it's divided by 4 and just wanted to see if there were any Illumina docs on this being the correct usage.

Reply:
Hi @Emiller88,

The help message states this
```
--bcl-num-parallel-tiles arg           # of tiles to process in parallel (default 1)
--bcl-num-conversion-threads arg       # of threads for conversion (per tile, default # cpu threads)
--bcl-num-compression-threads arg      # of threads for fastq.gz output compression (per tile, default # cpu threads,
                                       or HW+12)
--bcl-num-decompression-threads arg    # of threads for bcl/cbcl input decompression (per tile, default half # cpu 
                                       threads, or HW+8. Only applies when preloading files)

```

It seems to me, `bcl-convert` is made to take over the whole node it runs on since it's meant to be container native (for usage on the dragen platform). I think it just assumes to use all cores available in to the container.

We could set `${task.cpus}` to `bcl-num-parallel-tiles` and all the rest to 1, but I think that would be detrimental to users on a slower FS. The exact configuration is very dependent on the type of run and the hardware you run it on.

TL;DR, the current config is more 'just to be safe' and not very performance tuned. If it were up to me, I'd revert this and set everything to `${task.cpus}`. We've had no issues so far with such a setting.

As allways, open to suggestions!

Reply:
> the current config is more 'just to be safe' and not very performance tuned

This was the piece I was afraid of. Maybe a quick run with both options? It also looks like you could just set `--bcl-num-parallel-tiles ${task.cpus}` and let `bcl-convert` figure out how it wants to use those threads.

Reply:
Actually, I agree. The true parallelisation comes from the parallel handling of the tiles. We could argue the last 3 args are purely for performance tweaking.

How about we drop the latter 3 args and let users start with the defaults, enabling them to tweak as they go using `ext.args`?

Reply:
Sounds good to me, I'm also happy to leave them/add them back in, I just think we should test it to make sure we're utilizing all of the resources.